### PR TITLE
Threaded solver

### PR DESCRIPTION
**Pull request by enricotenuti**
I added the threaded version of the solver in a new file called thread.h. The threaded search uses a "thread poll" approach, the queue of tasks is created by a BFS search at depth 2 and then it is populated as tasks for the actual IDA*.

### CODE REVIEW

**Comment by enricotenuti** [comment]
With the last commit I fixed some variables initialization, now there is a bug with the solutions vector when we use a higher number of threads.

**Comment by enricotenuti** [comment]
The solutions vector had no termination character in the end and with the test cases provided the single thread solver was finding a different optimal solution and it didn't manifest the bug, differently by the version with 64 threads that was finding a solution with fewer characters. This fix works also with a higher number of solutions in the vector.

File: src/solvers/h48/thread.h
Comment (sebastianotronto) [inline src/solvers/h48/thread.h]: I suggest renaming this to `solve_h48_multithread`.

File: src/solvers/h48/thread.h
Comment (sebastianotronto) [inline src/solvers/h48/thread.h]: Maybe a better name for this file is `solve_multithread.h`.

**Review by sebastianotronto — CHANGES_REQUESTED** [review]
After today's debugging session I have done a few more tests, and I am confident this works (save for the bug with 1-moves solutions; 2 movers seem to be fine).

As for performance, I have not tested it much but I found a factor of around 1.9 with 2 threads, which seems pretty good! I'll test with more threads when my PC is fixed.

I made only a couple of suggestions for renaming stuff. I think there are a few things that can be written in a simpler way, but I'll let you decide if and when to do it.

Before I merge this you have to
1. Fix the bug for 1-moves solutions. For example, `./run solve -solver h48 -options "0;4;20" -cube "$(./run frommoves -moves "F")"` returns an 8-mover as shortest solution.
2. Pull in the changes from master (there are couple of merge conflicts) and re-generate the h0k4 table for testing.
3. EDIT: and fix the data race with `nsols`

File: src/solvers/h48/thread.h
Comment (sebastianotronto) [inline src/solvers/h48/thread.h]: I get a data race here (with line 50). The variable nsols should me mutex-guarded or made atomic.

**Review by sebastianotronto — APPROVED** [review]
Nice!

**Pull request merged**
